### DISCUSSION TITLE
Optimistic Updates With GraphQL

### DISCUSSION THREAD

**Discussion by Undistraction** [reduxjs/redux-toolkit#3545] [Q&A] [answered]
I'm implementing optimistic updates for an app that is using a GraphQL API and I'm finding it to be much more involved than I was expecting which is making me think there has to be a better approach. The main problem is that in order to optimistically update the cache, I need to know not only each query that the update needs to patch, but also all the arguments for those queries.

For example, imagine I have a `getUser` query and that returns a user with some `posts` (used to render a list of posts). I then have a separate query for `getPost` (used to render an individual post). If I want to use optimistic updating for my `updatePost` mutation, I need to patch both the `getUser` query and the `getPost` query, and to do that I need the `id` of the user, plus any other params that were passed in to that query. So now I also need to pass the `userId` into `updatePost` so that it is available  inside `onQueryStarted`.  This means that a component deep in the hierarchy that should only need to care about an individual post now also needs to know about the user.

Additionally, having to 'know' about which queries to patch seems like something that is duplicating the use of tags to provide and invalidate data. 

What am I missing here?

**Comment by markerikson** [comment]
RTKQ works by serializing the arguments so it can generate the cache key, like `getPokemon('pikachu')`.

Hypothetically, how would you expect to be able to update individual cache entries without knowing the way to look them up?

**Reply by Undistraction** [reply]
I understand the why, and there is a nice purity to it, but in practical application the idea that every mutation also needs to know about every query that might need to be patched, and all the arguments they took seems seems pretty clunky. I don't see how to do that elegantly without prop-drilling. Some component deep in the DOM that needs to fire an update off now has to know about all kinds of data which it shouldn't have to care about so it can pass that data into the mutation function. 

I guess one option would be to pull the arguments from query in the store by using `getState` inside `onQueryStarted`. That would give access to the args without needing to pass them around. Would that make sense?

**Reply by markerikson** [reply]
Hmm. Maybe https://redux-toolkit.js.org/rtk-query/api/created-api/api-slice-utils#selectinvalidatedbywould help here?

**Reply by EskiMojo14** [reply]
you could also just pull all the cached entries for a given endpoint: 

```ts
const cachedEntries = Object.values(getState()[api.reducerPath].queries).filter((substate) => substate.endpointName === 'getPost')

const args = cachedEntries.map((entry) => entry.originalArgs);
```

(this does involve operating on the "internal" shape of RTKQ's slices, but i think the chances of this breaking in future are low)

**Reply by Undistraction** [reply]
@markerikson thanks for the suggestion, but not sure I follow as to how `selectInvalidatedBy` would help here.

@EskiMojo14 thanks. That's definitely interesting. Would be nice to have a `selectArgsForQuery` utility.

**Reply by Undistraction** [reply]
Thinking about this some more, I guess that:
- This is more of a problem for GraphQL APIs because often queries are more general and include chunks of object graph, meaning that the branches and leaves of the graph might need to be updated, but to do so requires the initial query params, for example:
```
user(id) {
  items {
    nodes {
      id
      status {
        id
        isChosen
      }
    }
  }
}
```
(And queries can get much deeper than this)

There might be an `updateStatus` mutation, that only needs a Status `id` param, but to patch the `user` query it also needs the user `id`.

- This is a problem that might be avoided with a normalized cache, because it's easy to find and patch a record of a particular type (not saying a normalized cache is better, just that it would avoid this issue).

**Reply by EskiMojo14** [reply]
> @EskiMojo14 thanks. That's definitely interesting. Would be nice to have a selectArgsForQuery utility.

Sure, it's something that comes up every once in a while - I've opened a PR with it https://github.com/reduxjs/redux-toolkit/pull/3547 and will see how the other maintainers feel.

**Reply by phryneas** [reply]
I'd say you are getting to the point where you might want to take a normalized GraphQL client like Apollo Client into account.
RTK Query is okay if you have very light usage of GraphQL and treat your API more like a Rest API - but if you really want to go into the benefits of GraphQL, you might be better off with a specialized client.

**Reply by Undistraction** [reply]
Yeah. I'm starting to come to that conclusion. It really doesn't feel like RTKQuery is suited for GraphQL at all.

**Reply by phryneas** [reply, marked as answer]
It can cache GraphQL requests, but we are very upfront that it is a document cache, not a normalized cache.   
At this rate, RTK Query is about the level of `urql` without their optional Graphcache - no normalization, the only good way of updating data is automatic invalidation. That's okay for smaller apps, but not for advanced use of GraphQL.
